engine/runtime/src/main/java/org/enso/interpreter/runtime/data/Ref.java

```diff
@@ -14,16 +17,24 @@
 @ExportLibrary(TypesLibrary.class)
 @Builtin(pkg = "mutable", stdlibName = "Standard.Base.Runtime.Ref.Ref")
 public final class Ref implements EnsoObject {
+  /**
+   * {@code 0} - regular reference to an object {@code 1} - reference via {@link SoftReference}
+   * {@code 2} - reference via {@link WeakReference}
+   */
+  private final byte type;
+
   private volatile Object value;
 
   /**
    * Creates a new reference.
    *
    * @param value the initial value to store in the reference.
+   * @param referenceType type of reference to use
    */
   @Builtin.Method(description = "Creates a new Ref", autoRegister = false)
-  public Ref(Object value) {
-    this.value = value;
+  public Ref(Object value, long referenceType) {
+    this.type = (byte) (referenceType & 0x03);
```

Comment:
Why do we want this `& 0x03` here?

IMHO if the `referenceType` is outside of the currently allowed set {0, 1, 2}, we should just throw an `IllegalArgumentException`, not clamp the value so that 12 magically becomes 0.

It doesn't 'sanitize' the arguments fully anyway - by `x & 0x03` you limit `x` to the set {0, 1, 2, 3}, so if the 3 is passed as input, we can _still_ get an `IllegalStateException` from `ReferencesManager.create` later on. So currently e.g. `12` will get clamped to `0` and work, but `15` will get clamped to `3` and still throw.

Let's either clamp the input in such a way that all values become valid, or **check** the input and disallow invalid values by throwing an exception, let's not mix the two approaches.

IMO it's much cleaner to check the input and disallow any value outside of the allowed range of {0, 1, 2}. If we ever want to extend the API and allow some other kind of reference, we can then allocate one of the higher numbers for it. But if we do clamping like now, we would be breaking compatibility if we introduce reference types associated with numbers 4 and greater.

Reply:
Even better: `type` can be enum, and you can then convert `long` to the enum via something like `fromOrdinal`.

Reply:
> can be enum

- `enum` just like the ones implemented by #7607
- I am afraid that's not going to work as this is a builtin, not `polyglot java import...`

Reply:
Gone.